### PR TITLE
RHOAIENG-62244,RHOAIENG-62245: Multimodal chat UX — image/document upload, inline rendering, and thinking panel

### PR DESCRIPTION
**Pull request by Lucifergene**
https://issues.redhat.com/browse/RHOAIENG-62244
https://issues.redhat.com/browse/RHOAIENG-62245

## Description

Implements the full multimodal chat UX for the gen-ai Playground, covering Stories 1.2 (image upload component) and 1.3 (inline rendering, thinking panel, document chips, and send-button defaults).

### Story 1.2: Image upload component
- **Attach menu** with Upload image, Upload audio (disabled), Upload documents options
- **Client-side validation** for file type (JPG/PNG) and size (10MB limit) with auto-dismissing alerts
- **XHR-based upload** to `/gen-ai/api/v1/lsd/files/vision` with progress tracking
- **Image preview chip** with loading spinner and close/remove button
- **One-image-per-conversation** enforcement with disabled menu item
- **Vision-specific constants** (`VISION_UPLOAD_CONFIG`)

### Story 1.3: Multimodal chat UX — inline rendering, thinking panel, and document chips
- **Inline image rendering** in chat transcript (300px max-height thumbnail, border-radius)
- **Custom StreamingThinkingSection** component using PatternFly `ExpandableSection`
- **Live streaming of reasoning tokens** with auto-collapse when main answer begins
- **BFF `<think>` tag parsing** via `ThinkTagParser` class — handles `<think>reasoning</think>answer`, bare `reasoning</think>answer`, and partial tags split across SSE chunks
- **`response.reasoning_text.delta`** event handling for dedicated reasoning channels
- **Inline document chips** — transient in input area (with upload/close), persistent read-only in sent messages
- **"Replace media file?" modal** when re-uploading an image
- **Always-show send button** when attachments (image/docs) are ready, even with empty text
- **Default message injection** — "Describe the image" / "Summarize the attached documents" / combined
- **Uppercase image extension normalization** (.PNG → .png for icon display)
- **Blob URL revocation** on unmount and new-chat to prevent memory leaks
- **Merged streaming completion** to eliminate one-frame flicker
- **thinkBuffer flush emits onStreamData** for live UI consistency

### CodeRabbit review fixes
- **XHR upload timeout** — 60s timeout + `ontimeout` handler to prevent indefinite hangs
- **Stale upload race guard** — generation counter prevents aborted upload callbacks from wiping replacement state
- **Clear doc chips on external reset** — `clearAllMessagesRef` now also clears `pendingDocChips`
- **Attachment chip key collision** — index-suffixed keys prevent DOM reuse for same-name files
- **reasoningContent typeof guard** — runtime check before rendering thinking collapsible
- **Bare `</think>` streaming** — initial parser state now handles `reasoning</think>answer` without `<think>` prefix
- **Wire `isAudioUploadDisabled`** — prop destructured and applied to audio menu item
- **Test timer safety** — `jest.useFakeTimers` wrapped in try-finally to prevent leaks

## How Has This Been Tested?

- **Unit tests**: All 1118 gen-ai frontend tests pass (82 suites)
- **Playwright E2E**: Full flow tested against live cluster:
  1. Image upload + inline preview + send with vision model
  2. Reasoning models with streaming thinking tokens
  3. Document chip upload + inline rendering in transcript
  4. Replace media modal confirm/cancel
  5. Empty-text send with attachments → default message injection

## Test Impact

**New tests added:**

`thinkTagParser.spec.ts` (29 tests, new file):
- 9 `describe` blocks covering all state machine paths: dedicated channel, no tags, full `<think>...</think>`, empty block, partial open tag, bare `</think>`, partial close tag, stream-ends-mid-think flush
- Zero `mockFetch` — directly instantiates `ThinkTagParser` and feeds deltas

`ChatbotPlayground.spec.tsx` (33 tests):
- Inline doc chip lifecycle (creation, status sync, removal, send clearing)
- Send button gating (disabled while uploading, enabled when uploaded)
- Default message injection (image-only, docs-only, combined, no-injection with text)
- `alwaysShowSendButton` prop propagation
- Replace media modal (open, confirm with revoke, cancel)

`llamaStackService.spec.ts` (slimmed to wiring smoke tests):
- Reasoning channel wiring (dedicated `reasoning_text.delta` → `onStreamData`, embedded `<think>` → split)
- Chunk-level edge cases moved to `thinkTagParser.spec.ts`

`useChatbotMessages.spec.ts` (multiple new tests):
- Multimodal input construction, image persistence, reasoning effort payload
- Inline image/doc rendering in messages

`ChatbotMessageInput.spec.tsx` (new tests):
- Document chip rendering, loading states, close buttons
- `alwaysShowSendButton` prop

`StreamingThinkingSection.spec.tsx` (new file):
- Expand/collapse behavior, auto-collapse on streaming end

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added
- [x] The code follows our [Best Practices](/docs/best-practices.md)

If you have UI changes:
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

File: packages/gen-ai/frontend/src/app/Chatbot/ChatbotPlayground.tsx [hunk 1]
```diff
@@ -437,6 +530,14 @@ const ChatbotPlayground: React.FC<ChatbotPlaygroundProps> = ({
...
         onConfirm={() => {
           messageHooksRef.current.forEach((hook) => hook.clearConversation());
+          setHasImageInConversation(false);
+          setImageUploadState({
+            status: 'idle',
+            progress: 0,
+            fileId: null,
+            previewUrl: null,
+            fileName: null,
+          });
```
[the remark this answers is not in the document]
Comment (Lucifergene) [reply packages/gen-ai/frontend/src/app/Chatbot/ChatbotPlayground.tsx]: Fixed — was using `status: 'idle'` instead of `uploading: false`. The `ImageUploadState` interface uses `uploading: boolean`, not a status string.

File: packages/gen-ai/frontend/src/app/Chatbot/components/ChatbotMessageInput.tsx [hunk 2]
```diff
@@ -1,72 +1,122 @@
...
-      onShowErrorAlert?.(
-        `${ERROR_MESSAGES.FILE_UPLOAD_REJECTED}: ${errorMessages}`,
-        'File Upload Error',
-      );
+  const handleMenuSelect = React.useCallback((action: string) => {
+    setIsAttachMenuOpen(false);
+    if (action === 'upload-image') {
+      imageInputRef.current?.click();
+    }
+  }, []);
```
[the remark this answers is not in the document]
Comment (Lucifergene) [reply packages/gen-ai/frontend/src/app/Chatbot/components/ChatbotMessageInput.tsx:92]: Intentional — "Upload documents" and "Upload audio" are placeholders for future menu items. This PR is scoped to image upload. Document upload was previously handled through PF's native file dropzone path and hasn't been wired through the new custom attach menu yet.

File: packages/gen-ai/frontend/src/app/Chatbot/hooks/__tests__/useChatbotMessages.spec.ts [hunk 3]
```diff
@@ -674,4 +674,88 @@ describe('useChatbotMessages', () => {
...
+        ],
+      });
+      expect(secondPayload.chat_context![1]).toMatchObject({
+        role: 'assistant',
+        content: 'This is a bot response',
+      });
+
+      // Restore default getId behavior
+      mockGetId.mockImplementation(() => 'mock-id');
+    });
```
[the remark this answers is not in the document]
Comment (Lucifergene) [reply packages/gen-ai/frontend/src/app/Chatbot/hooks/__tests__/useChatbotMessages.spec.ts:769]: Fixed — wrapped the test body in try-finally so the `getId` mock is always restored even if assertions fail.

File: packages/gen-ai/frontend/src/app/services/llamaStackService.ts [hunk 4]
```diff
@@ -502,6 +502,45 @@ export const getFileUploadStatus = modArchRestGET<FileUploadStatusResponse>(
...
+          reject(new Error('Invalid response from server'));
+        }
+      } else {
+        reject(new Error(`Upload failed: ${xhr.status} ${xhr.statusText}`));
+      }
+    };
+    xhr.onerror = () => reject(new Error('Network error during upload'));
+    xhr.onabort = () => reject(new Error('Upload aborted'));
+    xhr.send(formData);
+  });
```
[the remark this answers is not in the document]
Comment (Lucifergene) [reply packages/gen-ai/frontend/src/app/services/llamaStackService.ts:603]: No CSRF token mechanism exists in the gen-ai frontend — zero matches for CSRF/csrf/csrfToken across the entire source. The BFF uses same-origin session cookies; the browser's same-origin policy handles this. Adding a non-existent `getCsrfToken()` call would be cargo-culting.

File: packages/gen-ai/frontend/src/app/services/llamaStackService.ts [hunk 5]
```diff
@@ -502,6 +502,45 @@ export const getFileUploadStatus = modArchRestGET<FileUploadStatusResponse>(
...
+        }
+      };
+    }
+
+    xhr.onload = () => {
+      if (xhr.status >= 200 && xhr.status < 300) {
+        try {
+          resolve(JSON.parse(xhr.responseText));
+        } catch {
+          reject(new Error('Invalid response from server'));
```
[the remark this answers is not in the document]
Comment (Lucifergene) [reply packages/gen-ai/frontend/src/app/services/llamaStackService.ts:593]: Fixed — added shape validation before resolving. The parsed response is now checked for `typeof parsed.data.id === 'string'` and rejects with `'Invalid response shape: missing data.id'` if malformed. Added a corresponding test case.

**Review by jharan1 — APPROVED** [review]
